[Chore] Update docs (#63)
serokell/tzbot: docs/development.md

and may be replaced with a better alternative later.

After this, the parser goes through the resulting tokens and tries to collect all time references remembering
their original text. The core element of a time reference is time of day, it must be present
in each one; some other optional elements can also be contained: location reference, date reference.
The ordering of all these elements can be arbitrary but each element can be found only once per time reference.

Following formats of time are supported:
* `7pm`
* `7 PM`
* `7.30pm`
* `7:30pm`
* `7:30`
* `19h`
* `19h30`

`7.30` is not supported intentionally because it rather represents a fractional number than time.

Current parser implementation is mainly focused on wide variety of different correct
variations and doesn't concern about possible grammatically incorrect, but successfully
parsed combinations.

Sometimes message text can have rather tricky structure when date reference can be away from the time itself:
  > Tomorrow, I'll be able to connect between 10am and 11am

This issue seems to require analyzing sentence structure and is quite subtle and interesting; we hope
to solve it one day :smile:

> ### Time translation
> 
> Main goal of the `TzBot.TimeReference` module is to try to translate the obtained `TimeReference` to the UTCTime
> collecting some additional meta info and handling possible timeshift errors. If date reference is not total, we
> try to guess what date was really meant (e.g. `chooseBestMonth` and `chooseBestYear`). In situations when exact date
> can't be defined, the real mentioned date can be slighly different and if a timeshift in sender/receiver's timezone
> is somewhere around then time conversion correctness can depend on the exact date. The translation module tries
> to pick some reasonable interval around the resulting time and check it for timeshifts. If one found, the user
> is warned about it.

Reviewer: Added more detail and examples to this section.

```suggestion
### Time translation

Main goal of the `TzBot.TimeReference` module is to try to translate the obtained `TimeReference` to the UTCTime
collecting some additional meta info and handling possible timeshift errors.

Sometimes the user provides incomplete information, and we have to infer what they meant.

* "1am": the user supplied the time, but not the date or the time zone.
  * In this case, we interpret "1am" as 1am of the current day; unless that time is already in the past,
    in which case we infer they must mean 1am the next day.
  * When the time zone is omitted, we infer that the reference must be relative to the sender's own time zone.
* "1am on wednesday": the user supplied the time and day of the week, but they didn't specify which week.
  * This is interpreted as 1am of the current day if today is wednesday.
    If it's not, then we infer it must mean 1am the next wednesday.
* "1am on the 12th": the user supplied the time and day of the month, but not the month or year.
  * See `chooseBestMonth` and `chooseBestYear`.
* "1am on March 12th": the user supplied the time, day of the month, and month, but not the year.
  * See `chooseBestYear`.

If our inferrence does not match the sender's true intention, and if the time zone rules for the day
we inferred differ from the rules for the day they meant, then there is possibility that the conversion will
not be correct.

Additionally, the translation module tries to pick some reasonable interval around the resulting time and
check it for timeshifts. If one found, the user is warned about it.
```

Reviewer: > Additionally, the translation module tries to pick some reasonable interval around the resulting time and
check it for timeshifts. If one found, the user is warned about it.

Okay, this feature (implemented in #58) is pretty complex, so let's document it carefully. In particular, we want to document the reason this feature exists. To do this, we should demonstrate the problem we were trying to solve in the first place, and how we decided to solve it, with examples.

Here's my suggestion, let's replace the paragraph above with its own subsection:

```
## Time shift warnings

When the user does not specify a date and we have to infer it, sometimes this inferred date may be
close to a time shift (a time when the clocks go forward or backward).
When this happens, an incorrect guess on our part is very likely to result in a misleading conversion.

For example, say it's 01:00 AM, 25 Mar 2023, and the user is in Europe/London.
At this point in time, this time zone observes UTC+0.

If they send a message containing "10am", we'll infer they mean "10am today, 25 March".
If a reader is in Europe/Moscow (which observes UTC+3 all year round), we'll convert "10am" to "13:00, 25 March 2023, in Europe/Moscow".

However, if our inference is wrong, and they actually meant "10am tomorrow", then this conversion is completely wrong.
On the morning of the 26th, at 01:00, Europe/London [transitions from UTC+0 to UTC+1](https://www.timeanddate.com/time/change/uk/london?year=2023),
which means the result of the conversion _should_ have been "12:00, 26 March 2023, in Europe/Moscow", not "13:00".

For this reason, when we infer a date, and if that date is "near" (3 days before or after) a time shift,
we warn the user that the accurancy of the conversion depends on whether we inferred the date correctly or not.

> "10am", 25 March 2023 in Europe/London

> 13:00, Saturday, 25 March 2023 in Europe/Moscow

> _Warning: We inferred that "10am" refers to 25 March 2023 in Europe/London and converted it to Europe/Moscow, but there is a time change near this date_:
>
>   • _At 01:00, 26 March 2023 in Europe/London, the clocks are turned forward 1 hour(s)_.
>
> _Beware that if this inference is not correct and the sender meant a different date, the conversion may not be accurate._

Note that we check if there is a time shift "near" the inferred date in both the "source" time zone and in each receiver's time zones.
If there if a time shift in the source time zone, every channel member will see a warning.
If there is a time shift in a receiver's time zone, only that receiver will see a warning.
```

### Processing messages from channels and direct messages

Most probable type of the bot work is processing channel messages and mailing them to the channel members.
Ephemeral that is sent to the author contains either time reference errors or translated time reference
that are _not_ in the same timezone as author's. This will help the sender to check that he hasn't mistaken
when translating his time to that timezone; i.e. if the sender lives in Helsinki, "9am UTC" will be returned
to him as "11am in Europe/Helsinki". If there are no such time refs and no errors, the author will receive

> no ephemeral message at all. Other users will get all time translations, possibly as "your timezone is the same"

Reviewer: ```suggestion
no ephemeral message at all. Other users will get all time translations, possibly as "You are in this timezone"
```

for those with matching timezones.

We could just not send anything to users that have the same timezone as the reference has, but decided to
just note that this time reference doesn't need to be translated. This is because absence of the translation
can be caused by several reasons:
* the bot doesn't work at all;
* it failed to parse this time;
* it wasn't able to send an ephemeral for some reason, etc.
By saying "your timezone is the same", we just say "everything is OK, but just see the original
time ref, it's correct for you".

Users can also DM the bot, in this case the bot will return him an ephemeral message such as other
users would see it if the message was posted in an ordinary channel.

AFAIU there is no way to add the bot to an existing DM, so messages in such chats can be only translated
using Slack entrypoints.

### Using entrypoints

> Currently there are two supported entrypoints in the message context menu: _Translate_ and _Report_.

Reviewer: ```suggestion
Currently there are two supported entrypoints in the message context menu: _Translate time refs_ and _Report an issue_.
```

The first one works just as expected:
it renders message text and translation pairs separately. The real advantage of this entrypoint is that it can
be triggered both for old messages for which all ephemerals are gone and for messages that are not visible

> for the bot by default (e.g. from direct message chats) The _Translate_ modal has "Report" button that leads
> to the _Report_ modal that is also reachable through mentioned _Report_ endpoint. There user can leave any
> feedback about the bot working wrongly.

Reviewer: ```suggestion
for the bot by default (e.g. from direct message chats). The _Translate time refs_ modal has a "Report" button that leads
to the _Report an issue_ modal. There, the user can leave
feedback about the bot working wrongly.
```

### Ignoring code blocks

The main reason for this is the fact that logs are often put into such blocks, and logs contain a lot of
timestamps, which can overload the bot server. Also code blocks often contain artificial character sequences
that can be wrongly recognized as time references, and we also want to avoid that.

This feature is implemented by analyzing Slack provided block structure of the incoming message
(see `TzBot.Slack.API.MessageBlock`). AFAIU it is not properly documented, the only link that describes
what's going on is [this API update report](https://api.slack.com/changelog/2019-09-what-they-see-is-what-you-get-and-more-and-less).
As the solution can't be considered stable, a reserve option is also added that tries to analyze code blocks manually.

### Caching data

For work the bot needs some data that changes relatively rarely:
* user info;
* conversation members;
* message details.
For this small cache wrapper over `cache` instance was implemented in `TzBot.Cache`.
Additionally, it supports expiry time randomization; it was initially introduced to avoid
multiple `user.info` calls at once. Also, as items are not automatically deleted after they

> get expired, a separate thread is running, purging all expired items form time to time.

Reviewer: ```suggestion
get expired, a separate thread is running, purging all expired items from time to time.
```
